File: #76_Minimum_Window_Substring/Solution.java
import java.util.*;

public class Solution {
    public String minWindow(String s, String t) {
    	String result = "";
    	if (s == null || s.length() == 0 || t == null || t.length() == 0 || s.length() < t.length())
			return result;    		
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        for (int i = 0; i < t.length(); i++) {
        	if (!map.containsKey(t.charAt(i))) map.put(t.charAt(i), 1);
        	else map.put(t.charAt(i), map.get(t.charAt(i)) + 1);
        }
        Map<Character, Integer> window = new HashMap<Character, Integer>();
        int i = 0, j = 0, minL = Integer.MAX_VALUE;
        int match = 0;
        while (i < s.length() && j < s.length()) {
        	if (map.containsKey(s.charAt(j))) {
        		if (window.containsKey(s.charAt(j))) 
        			window.put(s.charAt(j), window.get(s.charAt(j)) + 1);
        		else window.put(s.charAt(j), 1);
        		if (window.get(s.charAt(j)) <= map.get(s.charAt(j))) match++;
        	}
        	j++;
        	while (i < s.length() && match == t.length()) {
        		if (j - i < minL) {
        			minL = j - i;
        			result = s.substring(i, j);
        		}
        		if (!map.containsKey(s.charAt(i))) {
        			i++;
        			continue;
        		}
        		if (window.get(s.charAt(i)) >= map.get(s.charAt(i))) {
        			if (window.get(s.charAt(i)) == 1) window.remove(s.charAt(i));
    				else window.put(s.charAt(i), window.get(s.charAt(i)) - 1);
        		}
        		if (!window.containsKey(s.charAt(i)) || window.get(s.charAt(i)) < map.get(s.charAt(i))) match--;
    			i++;
        	}
        }
    	return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	// System.out.println(sol.minWindow("a", "aa"));
    	// System.out.println(sol.minWindow("aa", "aa"));
    	// System.out.println(sol.minWindow("ab", "a"));
		// System.out.println(sol.minWindow("ADOBECODEBANC", "ABC"));
		System.out.println(sol.minWindow("acbbaca", "aba"));
    }
}

File: #76_Minimum_Window_Substring/Solution.py
'''
1, when right bound meet a char in t, add it to window, always make sure that sliding window is valid,
2, then shrink left side as much as possible
3, compare length
'''

class Solution:
	# @param {string} s
	# @param {string} t
	# @return {string}
	def minWindow(self, s, t):
		if not t or not s:
			return ''
		map = {}
		for c in t:
			map[c] = map.get(c, 0) + 1
		window = {}
		i, j, match = 0, 0, 0
		minL = len(s)
		ans = ''
		while j < len(s):
			if s[j] not in map:
				j += 1    # extend right bound
				continue
			window[s[j]] = window.get(s[j], 0) + 1 # add it to window
			if window[s[j]] <= map[s[j]]:
				match += 1
			j += 1
			while i < j:  # shrink left bound
				if s[i] not in map:
					i += 1
				elif window[s[i]] > map.get(s[i]):
					window[s[i]] = window[s[i]] - 1
					i += 1
				else:
					break
			if match == len(t):  # see if it's a match
				l = j - i
				if l <= minL: #
					minL = l  #
					ans = s[i:j]
		return ans

sol = Solution()
s = "ABBC"
t = "ABC"
print sol.minWindow(s, t)